src/Aspire.Hosting/Dashboard/ResourceLogSource.cs

```diff
@@ -0,0 +1,49 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Threading.Channels;
+using Aspire.Hosting.Dcp;
+using Aspire.Hosting.Dcp.Model;
+
+namespace Aspire.Hosting.Dashboard;
+
+internal sealed class ResourceLogSource<R>(IKubernetesService kubernetesService, R resource):
+    IAsyncEnumerable<IReadOnlyList<(string Content, bool IsErrorMessage)>>
+    where R: CustomResource
+{
+    public async IAsyncEnumerator<IReadOnlyList<(string Content, bool IsErrorMessage)>> GetAsyncEnumerator(CancellationToken cancellationToken)
+    {
+        if (!cancellationToken.CanBeCanceled)
+        {
+            throw new ArgumentException("Cancellation token must be cancellable in order to prevent leaking resources.", nameof(cancellationToken));
+        }
+
+        Stream stdoutStream = await kubernetesService.GetLogStreamAsync(resource, Logs.StreamTypeStdOut, follow: true, cancellationToken).ConfigureAwait(false);
+        Stream stderrStream = await kubernetesService.GetLogStreamAsync(resource, Logs.StreamTypeStdErr, follow: true, cancellationToken).ConfigureAwait(false);
+
+        var channel = Channel.CreateUnbounded<(string Content, bool IsErrorMessage)>(
+            new UnboundedChannelOptions { AllowSynchronousContinuations = false, SingleReader = true, SingleWriter = false });
+
+        _ = Task.Run(() => streamLogs(stdoutStream, isError: false), cancellationToken);
+        _ = Task.Run(() => streamLogs(stderrStream, isError: true), cancellationToken);
```

Comment:
I worry about what happens if there is an error reading from the stream. It looks like the error won't ever be logged and the channel will just stop getting new values to write.

I think an ILogger needs to be present here. If there is an error in stream logs (other the OCE which we can ignore) then write the error to it.

Also, should the channel be completed with an error if there is an exception? IMO yes, so the UI attempts to get working streams.

Need to unit test this as well.

Reply:
Added the logger and the logic to complete the channel with error if there is an exception. As for the unit tests, this will have to be addressed in the next PR I am afraid.